### DISCUSSION TITLE
🚀 Announcement:  Migration to Poetry for Dependency Management 🚀

### DISCUSSION THREAD

**Discussion by Pouyanpi** [NVIDIA-NeMo/Guardrails#838] [Announcements]
We’re excited to announce that NeMo Guardrails is transitioning to use [Poetry](https://python-poetry.org/) for dependency management! This change is part of our ongoing efforts to streamline development processes and enhance project stability and maintenance.

🔧 Current Status:
The migration is currently on a [feature branch ](https://github.com/NVIDIA/NeMo-Guardrails/tree/feature/poetry)and hasn’t been merged into the develop branch yet. Indeed, it does not only cover dependency management but an effort to bring overall enhancements to our development workflow. We invite all contributors to try out this new setup and adjust their local development environment accordingly to be prepared for when the changes go live (expect it to happen mid November 🗓️)

📝 Getting Started:
To help you with the transition, we’ve updated the [contribution guide](https://github.com/NVIDIA/NeMo-Guardrails/blob/feature/poetry/CONTRIBUTING.md) with detailed instructions. Be sure to check it out as you set up your environment.

💡 Feedback Matters:
We value your insights and encourage you to share any issues or suggestions in this thread. Your feedback will help us ensure a smooth transition for everyone involved.

👉 Note:
This update primarily impacts contributors work on NeMo Guardrails. End users will not experience any changes in functionality.

Thank you for your support and contributions to NeMo Guardrails!

**Comment by mdambski** [comment]
@Pouyanpi, thank you for your contributions. 

Reaching out with a question on poetry migration.

We noticed that the [PR](https://github.com/NVIDIA/NeMo-Guardrails/pull/826) with migration to Poetry changed the constraints for the `httpx` library. It was modified from `httpx>=0.24.1` to `httpx = "^0.24.1"`. This effectively disallows the installation of `httpx` in versions 1.0 and above, which were previously allowed. Was this intentional? In the future, this could conflict with some library updates or CVE fixes.

**Reply by Pouyanpi** [reply]
Hi @mdambski , yes just merged it #939
